Variable Flow and Electric Low Temperature Radiant System Setpoint Definition Enhancement

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8045 
 - This pull request is to implement the ability of variable flow and electric radiant systems to adjust how the actual setpoint is interpreted within EnergyPlus.  Previously, when using a throttling range, EnergyPlus interpreted the setpoint in the middle of the throttling range.  For example, in a heating system, if your throttling range was 2C and your setpoint was 20C, then the system would respond with half flow/power when the temperature to be controlled upon was 20C, full flow/power when the temperature was 19C, and zero flow/power when the temperature reached 21C.  Based on information from the Center for the Built Environment (CBE) at the University of California-Berkeley, most modern system control actual considers the setpoint as the zero flow point.  In other words, using the example above, 20C would be used as the zero flow point and the response would vary linearly until reaching full flow/power at 18C.  This work implements a new setpoint control strategy that allows the more modern as well as the existing EnergyPlus method to be selected by the user.  This involves a new input parameter for both of these low temperature radiant systems.  The default which is used for all existing input files is considered "HalfFlowPower", but the user can switch to "ZeroFlowPower".  The new input parameter, a new subroutine to interpret the input, a new subroutine to set the control temperature based on this and other user input, a unit test of the new "set" routine, modified documentation, and transition rules are all part of this pull request.

In addition, three input files were changed to exercise this new feature.  These files include: RadLoTempElecTermReheat.idf, RadLoTempElecTermReheatCondFD.idf, and RadLoTempHydrCtrlOpt.idf.  The results for these files WILL change significantly as each of these files has one radiant system that changes from the default HalfFlowPower to the new ZeroFlowPower to demonstrate that it is working.  Modified output files were checked and the results indicated that the new setpoint control type was working as expected. 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: NewFeature
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions (see above)
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Implementation of Setpoint Control

This commit includes the implementation of setpoint control for variable flow and electric low tempeature radiant systems.  This includes the code changes, new subroutines, modified IDD, and modified IDFs in the test suite.

RKStrand committed:
Setpoint Control Type Unit Test, Docs, and Transition

This commit includes the unit test for the new subroutine that determines the control temperature from various input parameters including the new setpoint control type.  It also includes documentation and transition files.

RKStrand committed:
Merge branch 'develop' into 8045-LowTempRadSysSetptInterpretEnhancement

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -44025,7 +44025,13 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
         \default MeanAirTemperature
-   A6 , \field Heating Design Capacity Method
+   A6,  \field Setpoint Control Type
+        \note (How setpoint temperature is defined)
```
I would take out the parentheses.  The use of a `\note` sorta implies that this is just a footnote of added info anyway.

RKStrand replied:
They are gone now.

Myoldmopar replied:
:+1:

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -44025,7 +44025,13 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
+        \note (How setpoint temperature is defined)
+        \type choice
+        \key HalfFlowPower
+        \key ZeroFlowPower
```
Hmm, I'll have to fully grok the use of these to determine if the naming is good.  It probably is, but they seem a little funny at first.

RKStrand replied:
Well, it's trying to deal with flow or power so that we can handle both the same way.  So, it's consistent between the variable flow and electric (both use the same key words).  Perhaps there's a better way, but I felt like this was reasonable.  If you feel there is something better, I can change it.

Myoldmopar replied:
OK, if those keywords are in place elsewhere, then let's call that precedent and leave them.  If they need to be adjusted, it can happen on both objects at a later time.  But probably not.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -44025,7 +44025,13 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
         \key SurfaceInteriorTemperature
         \default MeanAirTemperature
-   A6 , \field Heating Design Capacity Method
+   A6,  \field Setpoint Control Type
```
Do you need to alter the `\min-fields` for this object because of your new field?

RKStrand replied:
Fixed this--you are correct there were two of these that needed to be updated.

Myoldmopar replied:
:+1:

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -87,6 +87,11 @@ namespace LowTempRadiantSystem {
+    enum class LowTempRadiantSetpointTypes {
+      halfFlowPower,        // Controls system where the setpoint is at the 50% flow/power point
+      zeroFlowPower,        // Controls system where the setpoint is at the 0% flow/power point
+    };
```
Yay for enum class!

RKStrand replied:
😀

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -145,22 +150,30 @@ namespace LowTempRadiantSystem {
         // Default Constructor
             RadiantSystemBaseData()
-                : SchedPtr(0), ZonePtr(0), NumOfSurfaces(0), TotalSurfaceArea(0.0), ControlType(LowTempRadiantControlTypes::MATControl)
+        : SchedPtr(0), ZonePtr(0), NumOfSurfaces(0), TotalSurfaceArea(0.0), ControlType(LowTempRadiantControlTypes::MATControl),
```
Something looks funny with the indentation in this section.

RKStrand replied:
Fixed this now.

Myoldmopar commented on src/Transition/CreateNewIDFUsingRulesV9_4_0.f90:
```diff
@@ -422,6 +422,22 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+                  OutArgs(1:9)=InArgs(1:9)
+                  OutArgs(10) = 'HalfFlowPower'
+                  OutArgs(11:CurArgs+1)=InArgs(10:CurArgs)
+                  CurArgs = CurArgs + 1
```
These changes look fine from visual inspection.

Myoldmopar commented on tst/EnergyPlus/unit/LowTempRadiantSystem.unit.cc:
```diff
@@ -2201,3 +2204,64 @@ TEST_F(LowTempRadiantSystemTest, calculateOperationalFractionTest)
+    HydrRadSys(1).SetpointType = LowTempRadiantSetpointTypes::halfFlowPower;
+    expectedResult = 1.25;
+    actualResult = HydrRadSys(1).setOffTemperatureLowTemperatureRadiantSystem(scheduleIndex,throttlingRange);
+    EXPECT_NEAR(expectedResult, actualResult, acceptibleError);
```
:+1:

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1615,6 +1621,22 @@ namespace LowTempRadiantSystem {
+            return LowTempRadiantSetpointTypes::halfFlowPower;
+        } else if (UtilityRoutines::SameString(controlInput, "ZeroFlowPower")) {
+            return LowTempRadiantSetpointTypes::zeroFlowPower;
+        } else {
```
Can't actually get here based on how you have written the IDD changes, but it's fine.

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -5082,6 +5099,21 @@ namespace LowTempRadiantSystem {
+                return scheduleValue + 0.5 * throttlingRange; // hush the compiler
+        }
+        
+    }
```
:+1:  for base class methods!

Myoldmopar reviewed (commented):
This looks like a great set of changes.  The functional changes are pretty small, and they make sense.  There is at least one minor change to make in the IDD.

RKStrand committed:
Correct merge problem and fix review issue

Seems like the merge with develop introduced two errors.  These are now fixed.  Also addressed an issue brought up in the review.

RKStrand committed:
Merge branch 'develop' into 8045-LowTempRadSysSetptInterpretEnhancement

RKStrand committed:
Updated min fields in IDD

Forgot one other item from the review--fixed the min fields required for the two objects in the IDD that had something added.

RKStrand committed:
Fixed some spacing issues

Another previously missed review comment

RKStrand commented:
@Myoldmopar Looks like from the last commit I made yesterday that the bug that I introduced during the merge conflict got taken care of and things are back in order.  I made another merge from develop and fixed two other issues this morning.  I'm assuming CI will come back "clean" (just 3 files with diffs--the ones that we expect).  Hopefully this is ready for another round of review at that point?

rraustad commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -145,25 +150,33 @@ namespace LowTempRadiantSystem {
+        
         virtual void calculateLowTemperatureRadiantSystem(ZoneTempPredictorCorrectorData &dataZoneTempPredictorCorrector, Real64 &LoadMet) = 0;
-                
+                        
```
Did you catch that this blank line is now filled with spaces?

RKStrand replied:
@rraustad Whoops--I had missed that.  Nice catch--thanks!  Just fixed it and made another commit.

RKStrand committed:
Removal of strange blank spaces

Title says it all.

Myoldmopar reviewed (approved):
Looks good to me.

Myoldmopar commented:
I'll build locally and test out your transition changes, etc., but I think it'll be good to go.

RKStrand commented:
@Myoldmopar 👍

Myoldmopar commented:
Tested transition locally, all went fine.  Then ran the transitioned file and it also ran fine.  I think this is good to go in, but I'll wait until CI wraps up the last commit.  If it is clean (back to the 3 expected diffs), then I'll merge.

RKStrand commented:
Awesome--thanks, @Myoldmopar!

Myoldmopar commented:
The 
- 21 tests had: AUD diffs.
- 17 tests had: RDD diffs.

are all from the PR I merged earlier this morning.  The rest of them appear to be as expected.  Can you confirm?

RKStrand commented:
@Myoldmopar Yup--this looks as expected.  I think we are good to go.
